### DISCUSSION TITLE
Restore and/or downgrade ESP32 Dev Kit to Older/Alternate Firmware after safeboot partitioning

### DISCUSSION THREAD

**Discussion by happy2rv** [arendst/Tasmota#20318] [Q&A]
How can you restore factory partition layout to ESP32 after installing Tasmota 13+ with safeboot partition structure?  I installed the latest Tasmota to a devkit and I'm having a few issues with it like retaining MQTT settings.  I thought I'd install a previous version, so I've made multiple attempts to erase the flash using esptool, and install various Tasmota 12&13 releases using the web installer and esptool with and without erase, but it always restores 13.3.0.1 from safeboot.  Attempts to "upgrade" the firmware from the Firmware Upgrade menu just indicate switching to safeboot partition and return to the 13.3.0.1 build main menu.  I know downgrading to a previous Tasmota version isn't supported, but I'm not trying to retain anything.  How do I get rid of Tasmota altogether, for instance if I wanted to return to factory or custom Arduino IDE code?

**Comment by sfromis** [comment]
Not clear what you specifically did, but it is **always** possible to do a "cold" erase/flash using `esptool.py` if you do it right. There is no way for a Tasmota setup to block that.

Beware of the risk that you are acting on a different device than you think, like if having more than one device connected via serial, and/or is looking at another host name or IP address.

**Reply by happy2rv** [reply]
I think there may be some issue with this particular dev kit.  It flashed the latest Tasmota, but won't retain MQTT settings on reboot.  I wasnt sure if esptool was limited in which partitions it erased when doing an erase.  It appears to run erase flash rather quickly but does not give any errors.  Likewise attempts to flash it with any other firmware appear to flash correctly bit revert to Tasmota on reboot.

**Reply by sfromis** [reply]
Better triple check that you are not mixing up devices.

You cannot erase/flash something over Tasmota, and still have Tasmota after that. And do make sure that the flashing you do is not just something partial.

**Comment by jaburges** [comment]
I'm actually looking to do the same
New plug comes with Tasmota (with safeboot partition)
I would like to flash a Tasmota image without the safeboot partition as my ultimate goal is to flash ESPresense firmware (ESP32-C3).

I have read a bunch of stuff, but my knowledge of ESP partitions is weak. Is there a way (has to be OTA as these plugs are glued closed) to upload ESPresense firmware binary to the ESP32 chip. It has a 16MB flash_size, and I can successfully adjust the partitions - but i'm not sure what the partition size should be and where I should upload the binary. 
With safeboot activated I'm assuming that is the OTA_0 partition. Do I overwrite that or write the binary file to APP_0 partition?

**Reply by sfromis** [reply]
Tasmota has no procedures for downgrade, removing the safeboot partition scheme, or accommodation for flashing incompatible firmware. Thus, it would be a non-minor task to create a OTA tool to repartition and flash something else, maybe with "inspiration" from the Tasmota tools for Partition Wizard and Partition Manager.
